tests/test_pytorch/test_dynamic_shapes.py

```diff
@@ -0,0 +1,138 @@
+# Copyright The Lightning AI team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import os
+
+import pytest
+import torch
+import torch.multiprocessing as mp
+from lightning_habana import HPUAccelerator, HPUParallelStrategy, SingleHPUStrategy
+from lightning_habana.utils.resources import device_count
+from lightning_utilities import module_available
+
+if module_available("lightning"):
+    from lightning.pytorch import Trainer, seed_everything
+    from lightning.pytorch.demos.boring_classes import BoringDataModule, BoringModel
+elif module_available("pytorch_lightning"):
+    from pytorch_lightning import Trainer, seed_everything
+    from pytorch_lightning.demos.boring_classes import BoringDataModule, BoringModel
+
+
+class DynamicOpsBoringModel(BoringModel):
+    """Boring model with dynamicity."""
+
+    def __init__(self):
+        super().__init__()
+        self.layer = torch.nn.Linear(32, 16)
+
+    def forward(self, x):
+        x = super().forward(x)
+        # Dynamic op: boolean indexing
+        x = x[x < 0]
+        x = x.view(-1, 1)
+        return torch.mm(x, torch.eye(x.shape[1], device=x.device, dtype=x.dtype))
+
+    def training_step(self, batch, batch_idx):
+        loss = super().training_step(batch, batch_idx)
+        self.log("train_loss", loss["loss"])
+        return loss
+
+
+def run_training(tmpdir, hpus, model, data_module, metric_file_path, recipe_caching, return_dict):
+    """Init trainer and run fit."""
+    seed_everything(42)
+    _strategy = SingleHPUStrategy()
+    if hpus > 1:
+        _strategy = (
+            HPUParallelStrategy(start_method="spawn")
+            if not recipe_caching
+            else HPUParallelStrategy(
+                recipe_cache_path=os.path.join(tmpdir, "recipes"), recipe_cache_size=1024, start_method="spawn"
```

Comment:
do we support other methods ?

Reply:
other methods also work. This is only part of WA for hangs due to ddp and pytest